client/js/userlist.js

```diff
@@ -0,0 +1,108 @@
+"use strict";
+
+const $ = require("jquery");
+const fuzzy = require("fuzzy");
+const Mousetrap = require("mousetrap");
+
+const templates = require("../views");
+
+const chat = $("#chat");
+
+chat.on("input", ".users .search", function() {
+	const value = $(this).val();
+	const parent = $(this).closest(".users");
+	const names = parent.find(".names-original");
+	const container = parent.find(".names-filtered");
+
+	// Input content has changed, reset the potential selection
+	parent.find(".user").removeClass("active");
+
+	if (!value.length) {
+		container.hide();
+		names.show();
+		return;
+	}
+
+	const fuzzyOptions = {
+		pre: "<b>",
+		post: "</b>",
+		extract: (el) => $(el).text(),
+	};
+
+	const result = fuzzy.filter(
+		value,
+		names.find(".user").toArray(),
+		fuzzyOptions
+	);
+
+	names.hide();
+	container.html(templates.user_filtered({matches: result})).show();
+});
+
+chat.on("mouseenter", ".users .user", function() {
+	$(this).addClass("active");
+});
+
+chat.on("mouseleave", ".users .user", function() {
+	// Reset any potential selection
+	$(".users .user").removeClass("active");
+});
+
+exports.handleKeybinds = function(input) {
+	Mousetrap(input.get(0)).bind(["up", "down"], (_e, key) => {
+		const userlists = input.closest(".users");
+		let userlist;
+
+		// If input field has content, use the filtered list instead
+		if (input.val().length) {
+			userlist = userlists.find(".names-filtered");
+		} else {
+			userlist = userlists.find(".names-original");
+		}
+
+		const users = userlist.find(".user");
+
+		// Find which item in the array of users is currently selected, if any.
+		// Returns -1 if none.
+		const activeIndex = users.toArray()
+			.findIndex((user) => user.classList.contains("active"));
```

Comment:
This is certainly not ideal, but I couldn't find a better way. I could try to improve now that it's committed and working if you prefer (also, suggestions welcome).

Reply:
Active would reset whenever the user list updates, right? All the user list updates are just full re-renders.

You could try using this: https://api.jquery.com/index/

Reply:
> Active would reset whenever the user list updates, right? All the user list updates are just full re-renders.

Good catch, I missed that. It's correctly handled now, highlight will carry on if there is a user list update 🎉
But that was suuuuuch a wild goose chase though, with all the cases, had to keep multiple browsers opened, etc. I _think_ I have all the cases right now, but it would be great if you could give it a quick test to see if I missed something.
I so wish there were ways to write test cases for that, it would have been a breeze to list and fix all the cases. The day we have browser tests, I am so happy.

> You could try using this: api.jquery.com/index

Yeah I tried like different ways with that and couldn't go anywhere with it. I'll give it a quick try again, but I'm not too confident. Would love a hand if you want to give it a shot though.

Reply:
Yeah so I'm trying with things like the following, but no dice:

```js
const activeIndex = users.find.index(".active");
const activeIndex = users.index(".user.active");
const activeIndex = userlist.index(".user.active");
const activeIndex = userlist.index(".active");
```

There must be a way, I'm just not seeing it...